src/ray/gcs/tables.h

```diff
@@ -337,26 +358,113 @@ class Table : private Log<ID, Data>,
   using Log<ID, Data>::command_type_;
   using Log<ID, Data>::GetRedisContext;
 
- private:
   int64_t num_adds_ = 0;
   int64_t num_lookups_ = 0;
 };
 
-class ObjectTable : public Log<ObjectID, ObjectTableData> {
+template <typename ID, typename Data>
+class SetInterface {
+ public:
+  using DataT = typename Data::NativeTableType;
+  using WriteCallback = typename Log<ID, Data>::WriteCallback;
+  virtual Status Add(const JobID &job_id, const ID &id, std::shared_ptr<DataT> &data,
+                     const WriteCallback &done) = 0;
+  virtual Status Remove(const JobID &job_id, const ID &id, std::shared_ptr<DataT> &data,
+                        const WriteCallback &done) = 0;
+  virtual ~SetInterface(){};
+};
+
+/// \class Set
+///
+/// A GCS table where every entry is an addable & removable set. This class is not
+/// meant to be used directly. All set classes should derive from this class
+/// and override the prefix_ member with a unique prefix for that set, and the
+/// pubsub_channel_ member if pubsub is required.
+///
+/// Example tables backed by Set:
+///   ObjectTable: Stores a set of which clients have added an object.
+template <typename ID, typename Data>
+class Set : private Log<ID, Data>,
+            public SetInterface<ID, Data>,
+            virtual public PubsubInterface<ID> {
+ public:
+  using DataT = typename Log<ID, Data>::DataT;
+  using Callback = typename Log<ID, Data>::Callback;
+  using WriteCallback = typename Log<ID, Data>::WriteCallback;
+  using NotificationCallback = typename Log<ID, Data>::NotificationCallback;
+  using SubscriptionCallback = typename Log<ID, Data>::SubscriptionCallback;
+
+  Set(const std::vector<std::shared_ptr<RedisContext>> &contexts, AsyncGcsClient *client)
+      : Log<ID, Data>(contexts, client) {}
+
+  using Log<ID, Data>::RequestNotifications;
+  using Log<ID, Data>::CancelNotifications;
+
+  /// Add an entry to the set.
+  ///
+  /// \param job_id The ID of the job (= driver).
+  /// \param id The ID of the data that is added to the GCS.
+  /// \param data Data to add to the set.
+  /// \param done Callback that is called once the data has been written to the
+  /// GCS.
+  /// \return Status
+  Status Add(const JobID &job_id, const ID &id, std::shared_ptr<DataT> &data,
+             const WriteCallback &done);
+
+  /// Remove an entry from the set.
+  ///
+  /// \param job_id The ID of the job (= driver).
+  /// \param id The ID of the data that is removed from the GCS.
+  /// \param data Data to remove from the set.
+  /// \param done Callback that is called once the data has been written to the
+  /// GCS.
+  /// \return Status
+  Status Remove(const JobID &job_id, const ID &id, std::shared_ptr<DataT> &data,
+                const WriteCallback &done);
+
+  Status Lookup(const JobID &job_id, const ID &id, const Callback &lookup) {
+    return Log<ID, Data>::Lookup(job_id, id, lookup);
```

Comment:
You can define these with `using Log<ID, Data>::Lookup` (and group with the above `using` statements).

Reply:
Yep!